src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -185,6 +188,95 @@ namespace OpenMS
+    MSChromatogram::ConstIterator it_rtend_r = chromatogram.RTEnd(right);
+
+    peakShapeMetrics.start_time_at_5 = findRTAtPoint(it_rtbegin_l, it_rtend_apex, 0.05, true);
+    peakShapeMetrics.start_time_at_10 = findRTAtPoint(it_rtbegin_l, it_rtend_apex, 0.1, true);
+    peakShapeMetrics.start_time_at_50 = findRTAtPoint(it_rtbegin_l, it_rtend_apex, 0.5, true);
+    peakShapeMetrics.end_time_at_5 = findRTAtPoint(it_rtend_apex, it_rtend_r, 0.05, false);
+    peakShapeMetrics.end_time_at_10 = findRTAtPoint(it_rtend_apex, it_rtend_r, 0.1, false);
+    peakShapeMetrics.end_time_at_50 = findRTAtPoint(it_rtend_apex, it_rtend_r, 0.5, false);
+
+    for (auto it = it_rtbegin_l; it != it_rtend_r; ++it)
+    {
+      peakShapeMetrics.points_across_baseline++;
+      if (std::max(it->getIntensity() - background_height_, 0.0) >= 0.5 * peak_height_)
+      {
+        peakShapeMetrics.points_across_half_height++;
+      }
+    }
+
+    // peak widths
+    peakShapeMetrics.width_at_5 = peakShapeMetrics.end_time_at_5 - peakShapeMetrics.start_time_at_5;
+    peakShapeMetrics.width_at_10 = peakShapeMetrics.end_time_at_10 - peakShapeMetrics.start_time_at_10;
+    peakShapeMetrics.width_at_50 = peakShapeMetrics.end_time_at_50 - peakShapeMetrics.start_time_at_50;
+    peakShapeMetrics.total_width = right - left;
+    peakShapeMetrics.slope_of_baseline = end_intensity - start_intensity;
+    peakShapeMetrics.baseline_delta_2_height = peakShapeMetrics.slope_of_baseline / peak_height_;
+
+    // other
+    peakShapeMetrics.tailing_factor =
+      peakShapeMetrics.width_at_5 /
+      std::min(
+        peak_apex_rt_ - peakShapeMetrics.start_time_at_5,
+        peakShapeMetrics.end_time_at_5 - peak_apex_rt_
+      );
+
+    peakShapeMetrics.asymmetry_factor =
+      std::min(
+        peak_apex_rt_ - peakShapeMetrics.start_time_at_10,
+        peakShapeMetrics.end_time_at_10 - peak_apex_rt_
+      ) /
+      std::max(
+        peak_apex_rt_ - peakShapeMetrics.start_time_at_10,
+        peakShapeMetrics.end_time_at_10 - peak_apex_rt_
+      );
+
+    return peakShapeMetrics;
+  }
+
+  double PeakIntegrator::findRTAtPoint(
+    MSChromatogram::ConstIterator it_begin,
+    MSChromatogram::ConstIterator it_end,
+    const double perc,
+    const bool is_start_time
+  ) const
+  {
+    MSChromatogram::ConstIterator it = std::find_if(it_begin, it_end, [perc, this](ChromatogramPeak pt)
+    {
+      return pt.getIntensity() >= perc * peak_height_;
+    });
+    const double int_r = std::max(it->getIntensity() - background_height_, 0.0); //background-subtracted intensity
+    const double int_l = std::max((it-1)->getIntensity() - background_height_, 0.0); //background-subtracted intensity of the previous point
+    const double rt_r = it->getMZ();
+    const double rt_l = (it-1)->getMZ();
+    return rt_r - (int_r - int_l) * (rt_r - rt_l) / (is_start_time ? int_r - perc * peak_height_ : perc * peak_height_ - int_r);
```

Comment:
Do you mind adding a few comments to make it a little more clear the intent of this function is?